bokeh/server/server.py

```diff
@@ -27,9 +27,10 @@ def _create_hosts_whitelist(host_list, port):
     hosts = []
     for host in host_list:
         if '*' in host:
-            log.warning('Host wildcard %r can expose the application to HTTP '
-                        'host header attacks. Host wildcard should only be '
-                        'used for testing purpose.', host)
+            log.warning(
+                "Host wildcard %r will allow websocket connections originating "
+                "from multiple (or possibly all) hostnames or IPs. Use non-wildcard "
+                "values to restrict access explicitly", host)
```

Comment:
I believe the purpose is that if an attacker makes an evil site, it can open a websocket to a Bokeh app I've visited, which might be an authenticated app with some proprietary information. Opening a websocket isn't constrained by same-origin policy.

The websocket will be opened with all the cookies that the Bokeh app has previously set, including for example authentication cookies. So the evil site's JavaScript will be able to silently log in with my credentials and download the Bokeh document and then upload it back to the attacker. The auth cookies I mean here may not be from Bokeh itself, they could be implemented in the app or implemented in a proxy in front of the app.

The Origin header is set by the browser, though, so as long as my browser hasn't been compromised, if I visit the evil site and it opens a websocket to a Bokeh app, the Bokeh app should get an accurate Origin with the evil site's domain in it, and reject that websocket.

A custom http client can forge the Origin header, but that doesn't do an attacker any good because they also need the authenticated user's cookies. The attack is based on tricking the authenticated user into browsing to the evil site using their real browser that's logged in to the Bokeh app, and then the evil site silently in the background can attack the Bokeh app. In that attack scenario, the Origin header will be accurate.

I could be wrong, and I'm not sure this is the only danger of allowing all websockets. If you find the original PR where we added that origin-checking stuff it might tell us more.

The reason we allow configuring additional websocket origins is that it's OK to let trusted sites (like another app you wrote) to connect. Also of course the public URL of the Bokeh app may not be the same as the current machine's hostname.